### PR TITLE
[Crosswalk-Mini][Android] LZMA compression support

### PR DESCRIPTION
**Pull request by d0ngjun**
Use lzma compress libxwalkcore.so to reduce apk size.

### CODE REVIEW

**Comment by halton** [comment]
> BTW, assets can't be used in Android Library project, you probably need to use res/raw instead.

@d0ngjun is working on this.

> use_lzma should not be a compile flag. Compressed and Uncompressed binaries should be both contained in packaging tool. 
For xwalk_core_library, we probably need to create xwalk_core_library_compressed besides existing xwalk_core_library.

Good suggestion and we ever think about that. But that would introduce several flags to use lzma or not:
* Test shell build time flag (current use_lzma):  for XWalkCoreShell and XWalkEmbeededClientShell 
* make_apk flag( to be added): for app developer. Default can be yes.
* Embedding API usage: for ant or Eclipse (not sure how to do that, but definitely can archive)

Clearlly, that will bring complexity. I doubt the value of output of this effort spending. Can we assume the CrosswalkMini customer only want to lzma currently. And report a feature request for future work?

**Comment by wang16** [comment]
Another issue comes to me is that. We are not carrying native libraries anymore. Therefore, GooglePlay can't tell the APK is for arm or ia. We should think this through.

**Comment by halton** [comment]
> Another issue comes to me is that, we are not carrying native libraries anymore, Therefore, GooglePlay can't tell the APK is for arm or ia. We should think this through.

Are you talking about the shared mode code that @lincsoon offered? If so, I'd like to resolve that in master branch, this branch only care about the embeeded mode and embedding api.

**Comment by wang16** [comment]
> Are you talking about the shared mode code that @lincsoon offered? If so, I'd like to resolve that in master branch, this branch only care about the embeeded mode and embedding api.

Not about shared mode. For each app created with compressed libxwalkcore.so. It doesn't contain a native library from a normal application's perspective. Even if developers upload both apks for ARM and IA to PlayStore, playstore will take both apks as non-native-binary included apks. So it will always offer the one with the higher version code to end user regardless of CPU architecture match.

**Comment by d0ngjun** [comment]
@halton @wang16 @lincsoon please help review, thanks.

File: runtime/android/core/src/org/xwalk/core/XWalkActivity.java [hunk 1]
```diff
@@ -5,16 +5,720 @@
...
+        @Override
+        protected Void doInBackground(Void... params) {
+            try {
+                XWalkCoreWrapper.decompressXWalkLibrary();
+                // TODO: use publishProgress to update percentage.
+            } catch (Exception e) {
+                Log.w("XWalkActivity", "decompress library failed: " + e.getMessage());
+            }
+
+            return null;
```
Comment (lincsoon) [inline runtime/android/core/src/org/xwalk/core/XWalkActivity.java]: doInBackground should pass the result of decompressXWalkLibrary to onPostExecute by return statement

File: runtime/android/core_shell/src/org/xwalk/core/xwview/shell/XWalkViewShellActivity.java [hunk 2]
```diff
@@ -359,4 +324,132 @@ public void onXWalkViewCreated(XWalkView view) {
...
+        XWalkViewShellActivity mActivity;
+
+        StartupDecompressDialog(XWalkViewShellActivity activity) {
+            super(activity);
+            mActivity = activity;
+
+            OnClickListener positiveListener = new OnClickListener() {
+                @Override
+                public void onClick(DialogInterface dialog, int id) {
+                    mActivity.finish();
```
Comment (lincsoon) [inline runtime/android/core_shell/src/org/xwalk/core/xwview/shell/XWalkViewShellActivity.java]: The asyncTask is still running in background until completion even if the activity has been finished. It's necessary to call asyncTask.cancel(). And do not forget to check isCancelled() for every loop in doInBackground()

File: runtime/android/core/src/org/xwalk/core/XWalkCoreWrapper.java [hunk 3]
```diff
@@ -0,0 +1,250 @@
...
+        Method method = null;
+        XWalkCoreStatus status = XWalkCoreStatus.NOT_FOUND;
+
+        try {
+            clazz = mBridgeLoader.loadClass(BRIDGE_PACKAGE + "." + XWALK_VIEW_DELEGATE);
+            method = clazz.getMethod("loadXWalkLibrary", Context.class);
+            method.invoke(null, xwalkApp);
+
+            status = XWalkCoreStatus.MATCHED;
+        } catch (UnsatisfiedLinkError | InvocationTargetException e) {
```
Comment (lincsoon) [inline runtime/android/core/src/org/xwalk/core/XWalkCoreWrapper.java]: UnsatiffiedLinkError means native library is built for ARM but running on IA image. There's no need to decompress and should return false directly.

File: runtime/android/core_internal/src/org/xwalk/core/internal/XWalkViewDelegate.java [hunk 4]
```diff
@@ -79,24 +79,44 @@
...
-                System.load("/data/data/" + context.getPackageName() + "/lib/" + library);
+        if (sLibraryLoaded || (context == null)) return;
+
+        if (XWalkLibraryCompressed(context)) {
+            String lib = PathUtils.getDataDirectory(context.getApplicationContext());
+            if (lib != null) {
+                for (String library : MANDATORY_LIBRARIES) {
+                    System.load(lib + "/" + library);
+                }
+            }
```
Comment (lincsoon) [inline runtime/android/core_internal/src/org/xwalk/core/internal/XWalkViewDelegate.java]: If the so file is compressed, is it necessary to call System.load for each file? Why not just throw an exception?

File: runtime/android/core/src/org/xwalk/core/XWalkActivity.java [hunk 5]
```diff
@@ -5,16 +5,720 @@
...
+    private XWalkCoreWrapper mCoreWrapper;
+    private LinkedList<Object> mReservedObjects;
+    private LinkedList<ReflectMethod> mReservedMethods;
+
+    public enum XWalkLibraryStatus {
+        MATCHED,
+        NOT_FOUND,
+        NEWER_VERSION,
+        OLDER_VERSION,
+        COMPRESSED,
```
Comment (lincsoon) [inline runtime/android/core/src/org/xwalk/core/XWalkActivity.java]: The XWalkLibraryStatus filed means the result of jar version check. Jar version and whether native so is compressed are two different things. It's better to treat them differently

File: runtime/android/core/src/org/xwalk/core/XWalkCoreWrapper.java [hunk 6]
```diff
@@ -0,0 +1,250 @@
...
+
+    public static XWalkCoreWrapper getInstance() {
+        return instance;
+    }
+
+    public static void initEmbeddedMode(XWalkCoreListener coreListener) {
+        reset(coreListener, null);
+        check();
+        init();
+    }
```
Comment (lincsoon) [inline runtime/android/core/src/org/xwalk/core/XWalkCoreWrapper.java]: Why remove the assert statement

File: runtime/android/core/src/org/xwalk/core/XWalkCoreWrapper.java (hunk 3 above)
Comment (d0ngjun) [inline runtime/android/core/src/org/xwalk/core/XWalkCoreWrapper.java]: according to http://developer.android.com/reference/java/lang/Runtime.html#loadLibrary(java.lang.String), seems 'not found' could trigger this exception also.

File: runtime/android/core_internal/src/org/xwalk/core/internal/XWalkViewDelegate.java (hunk 4 above)
Comment (d0ngjun) [inline runtime/android/core_internal/src/org/xwalk/core/internal/XWalkViewDelegate.java]: This means so files are compressed, so loadLibrary method should load the decompressed library from getDataDirectory.

File: runtime/android/core/src/org/xwalk/core/XWalkCoreWrapper.java (hunk 3 above)
Comment (lincsoon) [inline runtime/android/core/src/org/xwalk/core/XWalkCoreWrapper.java]: So how about letting loadXWalkLibrary throw a different exception if ARM library encounters IA image? What's your opinion @wang16 ?

Comment (wang16) [inline runtime/android/core/src/org/xwalk/core/XWalkCoreWrapper.java]: OK to me.

**Comment by xingnan** [comment]
Hi guys, do we have the data about decompression time on different devices?

**Comment by halton** [comment]
@xingnan, welcome back. The decompression time various on different devices, but basically better CPU and more memory have small time. So far the HongMi is around ~20s, which is kind of not acceptable. We're improving it.

**Comment by xingnan** [comment]
@halton Great, thanks for information :)

**Comment by wang16** [comment]
Could you do a git rebase -i to merge the fixup into shared mode commit? LGTM after that.

**Comment by d0ngjun** [comment]
@wang16 since lincsoon does not provide a single commit for this bug fix, rebase means I need to resolve a lot of conflict here, so I just use this fixup to remind myself resolve it later.

**Comment by d0ngjun** [comment]
@wang16 please wait a few minutes for code rebase & test.

**Comment by wang16** [comment]
Seems latest binary is still 13M with this commit? @halton

**Comment by halton** [comment]
Yes, I aware of that, same with my local build. @d0ngjun is working on that.

**Comment by halton** [comment]
Seems it is a buggy: it was using non-lzma libxwalkcore.so when first-run ninja. But when run ninja second time, it will use lzma libxwalkcore.so.

**Pull request merged**